From d0c34d36a7dbbdfe15634a64b3358a63fb8345b4 Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Tue, 05 Jul 2022 17:41:54 +0300
Subject: [PATCH] [SW-94323] gaudi2: reset cold_rst recovery_flag in erom_recovery

If the recovery flag is set the arcboot executes the erom_recovery
without even trying to load ppboot image and this flag is set in
arcboot rom whenever it fails to load ppboot image. Therefore
even after successful firmware in erom recovery the arcboot rom is
loading erom recovery image hence recovery flag needs to be cleared
in erom to allow arcboot rom to load the ppboot image after firmware
update.

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: I8bc8a69bf97afb6350c50f7382bcb10adc48a7dd
---

diff --git a/app/ext_rom/erom_recovery.c b/app/ext_rom/erom_recovery.c
index 16aa192..039effd 100644
--- a/app/ext_rom/erom_recovery.c
+++ b/app/ext_rom/erom_recovery.c
@@ -87,6 +87,8 @@
 	device_init_i2c_s();
 }
 
+__weak void preboot_clear_recovery_flag(void) {}
+
 void erom_main_function(void)
 {
 	u64 ver_off = (uintptr_t)preboot_vers;
@@ -134,6 +136,8 @@
 
 	preboot_clear_wd_gpio();
 
+	preboot_clear_recovery_flag();
+
 	cpu_boot_sts_set(CPU_BOOT_STATUS_WAITING_FOR_BOOT_FIT);
 
 	preboot_puts("In recovery mode\n");
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index 0a6bfd7..d3aed0c 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -1088,6 +1088,16 @@
 	clr_flr_reset();
 }
 
+#define COLD_RST_FLOPS_RECOVERY_FLAG_MASK	0x1
+void preboot_clear_recovery_flag(void)
+{
+	struct gaudi2_cold_rst_data cold_rst_data;
+
+	cold_rst_data.data = regmap_readl(mmPSOC_GLOBAL_CONF_COLD_RST_FLOPS_2);
+	cold_rst_data.recovery_flag = 1;
+	regmap_writel(cold_rst_data.data, mmPSOC_GLOBAL_CONF_COLD_RST_FLOPS_2);
+}
+
 /*
  * As per VLSI team drain time out of 0x2000 corresponds to 16 micro seconds
  * and this value should be enough for AXI drain time out
diff --git a/include/preboot.h b/include/preboot.h
index be68257..9ddee95 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -257,6 +257,7 @@
 
 void hl_flr_report_err(u32 err_bit);
 void preboot_security_sp_sram_init(void);
+void preboot_clear_recovery_flag(void);
 
 #endif /* __ASSEMBLY__ */
 
